File: .skills/openclaw-skills/skills/jackyhwei/odd-articles/scripts/distribute/wechat-api.ts
/**
 * WeChat Official Account API client.
 * Pushes articles directly to drafts via the WeChat MP API,
 * bypassing Chrome CDP automation entirely.
 */

import { spawnSync } from 'node:child_process';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import type { Manifest } from './cdp-utils.ts';

// ─── Constants ───

const WECHAT_API_BASE = 'https://api.weixin.qq.com';
// Bun's TLS stack rejects WeChat API certs on POST multipart uploads;
// this option bypasses certificate verification for these API calls.
export const BUN_FETCH_OPTS = { tls: { rejectUnauthorized: false } } as const;
const MAX_RETRIES = 2;
const RETRY_DELAY_MS = 3000;
const CONFIG_DIR = path.join(os.homedir(), '.config', 'wechat-api');
const CONFIG_FILE = path.join(CONFIG_DIR, 'config.json');
const TOKEN_CACHE_FILE = path.join(CONFIG_DIR, 'token-cache.json');
const TOKEN_LIFETIME_MS = 2 * 60 * 60 * 1000; // 2 hours
const TOKEN_REFRESH_BUFFER_MS = 5 * 60 * 1000; // refresh 5 min early

const MD_TO_WECHAT_SCRIPT = process.env.MD_TO_WECHAT_SCRIPT || path.join(import.meta.dirname, '../../baoyu-post-to-wechat/scripts/md-to-wechat.ts');

// ─── Types ───

interface Credentials {
  appId: string;
  appSecret: string;
}

interface TokenCache {
  accessToken: string;
  expiresAt: number; // epoch ms
}

interface MdToWechatOutput {
  title: string;
  author?: string;
  summary?: string;
  htmlPath: string;
  contentImages: Array<{
    placeholder: string;
    localPath: string;
    originalPath: string;
  }>;
}

// ─── Retry helper ───

export async function fetchWithRetry(url: string, init?: RequestInit & { tls?: { rejectUnauthorized: boolean } }): Promise<Response> {
  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      return await fetch(url, init);
    } catch (err) {
      if (attempt < MAX_RETRIES) {
        const msg = err instanceof Error ? err.message : String(err);
        console.log(`  [wechat-api] Request failed (${msg}), retrying in ${RETRY_DELAY_MS / 1000}s... (${attempt + 1}/${MAX_RETRIES})`);
        await new Promise(r => setTimeout(r, RETRY_DELAY_MS));
      } else {
        throw err;
      }
    }
  }
  throw new Error('unreachable');
}

// ─── Credentials ───

export function loadCredentials(): Credentials | null {
  const appId = process.env.WECHAT_APPID?.trim();
  const appSecret = process.env.WECHAT_APPSECRET?.trim();
  if (appId && appSecret) {
    return { appId, appSecret };
  }

  if (fs.existsSync(CONFIG_FILE)) {
    try {
      const config = JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
      if (config.appId && config.appSecret) {
        return { appId: config.appId, appSecret: config.appSecret };
      }
    } catch {}
  }

  return null;
}

// ─── Access Token ───

function readTokenCache(): TokenCache | null {
  if (!fs.existsSync(TOKEN_CACHE_FILE)) return null;
  try {
    const cache = JSON.parse(fs.readFileSync(TOKEN_CACHE_FILE, 'utf-8')) as TokenCache;
    if (cache.accessToken && cache.expiresAt > Date.now() + TOKEN_REFRESH_BUFFER_MS) {
      return cache;
    }
  } catch {}
  return null;
}

function writeTokenCache(token: string, expiresIn: number): void {
  fs.mkdirSync(CONFIG_DIR, { recursive: true });
  const cache: TokenCache = {
    accessToken: token,
    expiresAt: Date.now() + expiresIn * 1000,
  };
  fs.writeFileSync(TOKEN_CACHE_FILE, JSON.stringify(cache, null, 2));
}

export async function getAccessToken(creds: Credentials): Promise<string> {
  const cached = readTokenCache();
  if (cached) return cached.accessToken;

  const url = `${WECHAT_API_BASE}/cgi-bin/token?grant_type=client_credential&appid=${encodeURIComponent(creds.appId)}&secret=${encodeURIComponent(creds.appSecret)}`;
  const res = await fetchWithRetry(url, { ...BUN_FETCH_OPTS });
  if (!res.ok) throw new Error(`Failed to get access_token: HTTP ${res.status}`);

  const data = await res.json() as { access_token?: string; expires_in?: number; errcode?: number; errmsg?: string };
  if (data.errcode && data.errcode !== 0) {
    throw new Error(`WeChat API error ${data.errcode}: ${data.errmsg}`);
  }
  if (!data.access_token || !data.expires_in) {
    throw new Error('Invalid access_token response');
  }

  writeTokenCache(data.access_token, data.expires_in);
  console.log('  [wechat-api] Access token obtained');
  return data.access_token;
}

// ─── Image Upload ───

export async function uploadContentImage(token: string, imagePath: string): Promise<string> {
  const fileData = fs.readFileSync(imagePath);
  const ext = path.extname(imagePath).slice(1) || 'png';
  const mimeType = ext === 'jpg' ? 'image/jpeg' : `image/${ext}`;
  const fileName = path.basename(imagePath);

  const form = new FormData();
  form.append('media', new Blob([fileData], { type: mimeType }), fileName);

  const url = `${WECHAT_API_BASE}/cgi-bin/media/uploadimg?access_token=${token}`;
  const res = await fetchWithRetry(url, { method: 'POST', body: form, ...BUN_FETCH_OPTS });
  if (!res.ok) throw new Error(`uploadimg failed: HTTP ${res.status}`);

  const data = await res.json() as { url?: string; errcode?: number; errmsg?: string };
  if (data.errcode && data.errcode !== 0) {
    throw new Error(`uploadimg error ${data.errcode}: ${data.errmsg}`);
  }
  if (!data.url) throw new Error('uploadimg: no URL returned');

  return data.url;
}

export async function uploadCoverImage(token: string, imagePath: string): Promise<string> {
  const fileData = fs.readFileSync(imagePath);
  const ext = path.extname(imagePath).slice(1) || 'png';
  const mimeType = ext === 'jpg' ? 'image/jpeg' : `image/${ext}`;
  const fileName = path.basename(imagePath);

  const form = new FormData();
  form.append('media', new Blob([fileData], { type: mimeType }), fileName);

  const url = `${WECHAT_API_BASE}/cgi-bin/material/add_material?access_token=${token}&type=image`;
  const res = await fetchWithRetry(url, { method: 'POST', body: form, ...BUN_FETCH_OPTS });
  if (!res.ok) throw new Error(`add_material failed: HTTP ${res.status}`);

  const data = await res.json() as { media_id?: string; errcode?: number; errmsg?: string };
  if (data.errcode && data.errcode !== 0) {
    throw new Error(`add_material error ${data.errcode}: ${data.errmsg}`);
  }
  if (!data.media_id) throw new Error('add_material: no media_id returned');

  return data.media_id;
}

// ─── Markdown Conversion ───

export function convertMarkdown(markdownPath: string): MdToWechatOutput {
  console.log('  [wechat-api] Converting markdown to HTML...');

  const scriptDir = path.dirname(MD_TO_WECHAT_SCRIPT);
  const result = spawnSync('npx', ['-y', 'bun', MD_TO_WECHAT_SCRIPT, markdownPath, '--theme', 'grace'], {
    cwd: scriptDir,
    timeout: 60_000,
    encoding: 'utf-8',
  });

  if (result.status !== 0) {
    const stderr = result.stderr?.trim() || '';
    throw new Error(`md-to-wechat failed (exit ${result.status})${stderr ? `: ${stderr}` : ''}`);
  }

  const stdout = result.stdout?.trim();
  if (!stdout) throw new Error('md-to-wechat produced no output');

  // The script outputs JSON to stdout
  const output = JSON.parse(stdout) as MdToWechatOutput;
  if (!output.htmlPath || !fs.existsSync(output.htmlPath)) {
    throw new Error(`HTML file not found: ${output.htmlPath}`);
  }

  return output;
}

// ─── HTML Processing ───

export function extractArticleContent(htmlPath: string, inlineCss = false): { content: string; styles: string } {
  const html = fs.readFileSync(htmlPath, 'utf-8');

  // Extract <style> blocks
  const styleMatches = html.match(/<style[^>]*>([\s\S]*?)<\/style>/gi) || [];
  const styles = styleMatches.map(s => {
    const inner = s.replace(/<\/?style[^>]*>/gi, '');
    return inner.trim();
  }).join('\n');

  // Priority: #output → .content → <body>

  // 1. Extract #output content (md-to-wechat output)
  const outputMatch = html.match(/<div[^>]*id=["']output["'][^>]*>([\s\S]*?)<\/div>\s*(?:<\/body>|<script|$)/i);
  if (outputMatch) {
    const content = stripTipElements(outputMatch[1].trim());
    return inlineCss ? { content: inlineCssStyles(content, styles), styles: '' } : { content, styles };
  }

  // 2. Extract .container div (md2wechat_formatter _preview.html uses .container)
  const containerMatch = html.match(/<div[^>]*class=["'][^"']*\bcontainer\b[^"']*["'][^>]*>([\s\S]*?)<\/div>\s*<\/body>/i);
  if (containerMatch) {
    const content = stripTipElements(containerMatch[1].trim());
    return inlineCss ? { content: inlineCssStyles(content, styles), styles: '' } : { content, styles };
  }

  // 2b. Extract .content container (md2wechat_formatter _preview.html)
  const contentMatch = html.match(/<div[^>]*class=["'][^"']*\bcontent\b[^"']*["'][^>]*>([\s\S]*?)<\/div>\s*(?:<\/body>|<script|$)/i);
  if (contentMatch) {
    const content = stripTipElements(contentMatch[1].trim());
    return inlineCss ? { content: inlineCssStyles(content, styles), styles: '' } : { content, styles };
  }

  // 3. Fallback: extract <body> content
  const bodyMatch = html.match(/<body[^>]*>([\s\S]*?)<\/body>/i);
  if (bodyMatch) {
    const content = stripTipElements(bodyMatch[1].trim());
    return inlineCss ? { content: inlineCssStyles(content, styles), styles: '' } : { content, styles };
  }

  throw new Error('Could not extract article content from HTML');
}

/** Remove .tip elements (e.g. "全选复制粘贴到公众号编辑器" hint bars) */
function stripTipElements(html: string): string {
  return html.replace(/<div[^>]*class=["'][^"']*\btip\b[^"']*["'][^>]*>[\s\S]*?<\/div>/gi, '').trim();
}

interface CssRule {
  selector: string;
  declarations: Map<string, string>;
}

function parseCssRules(css: string): CssRule[] {
  const rules: CssRule[] = [];
  const ruleBlocks = css.match(/[^{}]+\{[^{}]*\}/g) || [];
  for (const block of ruleBlocks) {
    const colonIdx = block.indexOf('{');
    if (colonIdx === -1) continue;
    const selector = block.slice(0, colonIdx).trim();
    const declarations = block.slice(colonIdx + 1, block.lastIndexOf('}')).trim();
    const declMap = new Map<string, string>();
    for (const decl of declarations.split(';')) {
      const colon = decl.indexOf(':');
      if (colon === -1) continue;
      const prop = decl.slice(0, colon).trim().toLowerCase();
      const value = decl.slice(colon + 1).trim();
      if (prop && value) declMap.set(prop, value);
    }
    if (declMap.size > 0) rules.push({ selector, declarations: declMap });
  }
  return rules;
}

function cssSelectorMatches(selector: string, tagName: string, classes: string[], id: string): boolean {
  const parts = selector.trim().split(/\s+/);
  let lastPart = parts[parts.length - 1];
  if (lastPart === '*') return true;
  const tagMatch = lastPart.match(/^([a-zA-Z][a-zA-Z0-9]*)/);
  if (tagMatch && tagMatch[1].toLowerCase() !== tagName.toLowerCase()) return false;
  const classMatches = [...lastPart.matchAll(/\.([a-zA-Z_-][a-zA-Z0-9_-]*)/g)].map(m => m[1]);
  for (const cls of classMatches) {
    if (!classes.includes(cls)) return false;
  }
  const idMatch = lastPart.match(/#([a-zA-Z_-][a-zA-Z0-9_-]*)/);
  if (idMatch && idMatch[1] !== id) return false;
  return true;
}

function inlineCss(html: string, rules: CssRule[]): string {
  return html.replace(/<([a-zA-Z][a-zA-Z0-9]*)\b([^>]*)>/g, (_match, tag, attrs) => {
    const existingStyleMatch = attrs.match(/style=["']([^"']*)["']/);
    // If element already has inline styles, preserve them (md-to-wechat already inlined)
    if (existingStyleMatch) {
      return `<${tag}${attrs}>`;
    }
    
    const classMatch = attrs.match(/class=["']([^"']*)["']/);
    const idMatch = attrs.match(/id=["']([^"']*)["']/);
    const classes = classMatch ? classMatch[1].split(/\s+/).filter(Boolean) : [];
    const id = idMatch ? idMatch[1] : '';
    const matchedStyles: string[] = [];
    for (const rule of rules) {
      if (cssSelectorMatches(rule.selector, tag, classes, id)) {
        for (const [prop, value] of rule.declarations) {
          matchedStyles.push(`${prop}:${value}`);
        }
      }
    }
    const newAttrs = attrs.replace(/\s*style=["'][^"']*["']/g, '');
    if (matchedStyles.length > 0) {
      return `<${tag}${newAttrs} style="${matchedStyles.join(';')}">`;
    }
    return `<${tag}${newAttrs}>`;
  });
}

export function inlineCssStyles(html: string, css: string): string {
  if (!css) return html;
  const rules = parseCssRules(css);
  return inlineCss(html, rules);
}

export function processHtmlWithImages(
  content: string,
  styles: string,
  imageMap: Map<string, string>,
): string {
  let processed = content;

  for (const [placeholder, cdnUrl] of imageMap) {
    processed = processed.replaceAll(
      placeholder,
      `<img src="${cdnUrl}" style="max-width: 100%; height: auto;" />`,
    );
  }

  if (styles) {
    processed = inlineCssStyles(processed, styles);
  }
  return processed;
}

// ─── Draft Creation ───

export async function createDraft(
  token: string,
  article: {
    title: string;
    content: string;
    author?: string;
    digest?: string;
    thumb_media_id: string;
  },
): Promise<{ media_id: string }> {
  const url = `${WECHAT_API_BASE}/cgi-bin/draft/add?access_token=${token}`;

  const body = {
    articles: [
      {
        title: article.title,
        author: article.author || '',
        digest: article.digest || '',
        content: article.content,
        thumb_media_id: article.thumb_media_id,
        content_source_url: '',
        need_open_comment: 0,
        only_fans_can_comment: 0,
      },
    ],
  };

  const res = await fetchWithRetry(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    ...BUN_FETCH_OPTS,
  });

  if (!res.ok) throw new Error(`draft/add failed: HTTP ${res.status}`);

  const data = await res.json() as { media_id?: string; errcode?: number; errmsg?: string };
  if (data.errcode && data.errcode !== 0) {
    throw new Error(`draft/add error ${data.errcode}: ${data.errmsg}`);
  }
  if (!data.media_id) throw new Error('draft/add: no media_id returned');

  return { media_id: data.media_id };
}

// ─── Main Orchestrator ───

export async function publishViaApi(manifest: Manifest): Promise<{ mediaId: string }> {
  const wechatData = manifest.outputs.wechat;
  if (!wechatData?.markdown) throw new Error('No wechat markdown in manifest');

  // 1. Load credentials
  const creds = loadCredentials();
  if (!creds) throw new Error('No WeChat API credentials configured');

  console.log('  [wechat-api] Starting API publish flow...');

  // 2. Get access token
  const token = await getAccessToken(creds);

  // 3. Convert markdown to HTML (unless pre-rendered HTML provided)
  let title: string;
  let author: string | undefined;
  let digest: string | undefined;
  let htmlContent: string;
  let styles = '';
  let contentImages: MdToWechatOutput['contentImages'] = [];

  if (wechatData.html && fs.existsSync(wechatData.html)) {
    // Use pre-rendered HTML — inline all CSS into element styles
    console.log('  [wechat-api] Using pre-rendered HTML');
    const extracted = extractArticleContent(wechatData.html, true);
    htmlContent = extracted.content;
    styles = extracted.styles;
    title = wechatData.title || manifest.title;
    author = wechatData.author;
    digest = wechatData.digest;
  } else {
    // Convert markdown
    const mdOutput = convertMarkdown(wechatData.markdown);
    const extracted = extractArticleContent(mdOutput.htmlPath, true);
    htmlContent = extracted.content;
    styles = extracted.styles;
    title = wechatData.title || mdOutput.title || manifest.title;
    author = wechatData.author || mdOutput.author;
    digest = wechatData.digest || mdOutput.summary;
    contentImages = mdOutput.contentImages || [];
  }

  // 4. Upload content images and build placeholder→URL map
  const imageMap = new Map<string, string>();
  if (contentImages.length > 0) {
    console.log(`  [wechat-api] Uploading ${contentImages.length} content image(s)...`);
    for (const img of contentImages) {
      if (!fs.existsSync(img.localPath)) {
        console.warn(`  [wechat-api] Image not found, skipping: ${img.localPath}`);
        continue;
      }
      try {
        const cdnUrl = await uploadContentImage(token, img.localPath);
        imageMap.set(img.placeholder, cdnUrl);
        console.log(`  [wechat-api]   ✓ ${path.basename(img.localPath)}`);
      } catch (err) {
        console.warn(`  [wechat-api]   ✗ ${path.basename(img.localPath)}: ${err instanceof Error ? err.message : err}`);
      }
    }
  }

  // 5. Process HTML with uploaded image URLs
  const finalContent = processHtmlWithImages(htmlContent, styles, imageMap);

  // 6. Upload cover image (required for draft)
  let thumbMediaId: string;
  const coverPath = wechatData.cover_image;

  if (coverPath && fs.existsSync(coverPath)) {
    console.log('  [wechat-api] Uploading cover image...');
    thumbMediaId = await uploadCoverImage(token, coverPath);
  } else if (wechatData.images && wechatData.images.length > 0 && fs.existsSync(wechatData.images[0])) {
    // Fallback: use first content image as cover
    console.log('  [wechat-api] No cover image specified, using first content image...');
    thumbMediaId = await uploadCoverImage(token, wechatData.images[0]);
  } else if (contentImages.length > 0 && fs.existsSync(contentImages[0].localPath)) {
    console.log('  [wechat-api] No cover image specified, using first article image...');
    thumbMediaId = await uploadCoverImage(token, contentImages[0].localPath);
  } else {
    throw new Error('No cover image available. Provide cover_image in manifest or ensure article has images.');
  }

  // 7. Create draft
  console.log('  [wechat-api] Creating draft...');
  const draft = await createDraft(token, {
    title,
    content: finalContent,
    author,
    digest,
    thumb_media_id: thumbMediaId,
  });

  console.log(`  [wechat-api] Draft created successfully (media_id: ${draft.media_id})`);
  return { mediaId: draft.media_id };
}

File: .skills/openclaw-skills/skills/279458179/wxgzh-mcp/src/wechat_api.py
"""WeChat Official Account API wrapper."""
import time
import requests
from typing import Optional, Dict, Any, List


class WeChatAPIError(Exception):
    """WeChat API error."""

    def __init__(self, message: str, code: int, raw: Dict = None):
        super().__init__(message)
        self.code = code
        self.raw = raw or {}


class WeChatAPI:
    """WeChat Official Account API client.

    Supports:
    - Access token management (auto-cached)
    - Image/thumb upload
    - Draft management
    - Material listing
    """

    BASE_URL = "https://api.weixin.qq.com/cgi-bin"

    def __init__(self, app_id: str = None, app_secret: str = None, proxy: str = None):
        """Initialize WeChat API client.

        Args:
            app_id: WeChat Official Account AppID
            app_secret: WeChat Official Account AppSecret
            proxy: HTTP proxy URL (optional)
        """
        from .config import get_config

        cfg = get_config()
        self.app_id = app_id or cfg.app_id
        self.app_secret = app_secret or cfg.app_secret
        self.proxy = proxy or cfg.proxy
        self._access_token: Optional[str] = None
        self._token_expires_at: float = 0

    def _get_proxies(self) -> Optional[Dict[str, str]]:
        """Get HTTP proxies dict."""
        if self.proxy:
            return {"http": self.proxy, "https": self.proxy}
        return None

    def _request(
        self,
        method: str,
        endpoint: str,
        params: Dict = None,
        json_data: Dict = None,
        data=None,
        timeout: int = 30,
    ) -> Dict[str, Any]:
        """Make HTTP request to WeChat API."""
        url = f"{self.BASE_URL}/{endpoint.lstrip('/')}"
        proxies = self._get_proxies()

        # Build headers
        headers = {"Content-Type": "application/json; charset=utf-8"}

        # If data is already encoded, use it directly
        if data is not None:
            resp = requests.request(
                method=method,
                url=url,
                params=params,
                data=data,
                proxies=proxies,
                timeout=timeout,
                headers=headers,
            )
        else:
            resp = requests.request(
                method=method,
                url=url,
                params=params,
                json=json_data,
                proxies=proxies,
                timeout=timeout,
                headers=headers,
            )

        resp.raise_for_status()

        # Handle both JSON and non-JSON responses
        content_type = resp.headers.get("Content-Type", "")
        if "application/json" in content_type:
            result = resp.json()
        else:
            result = {"raw": resp.text}

        # Check for WeChat API errors
        if isinstance(result, dict) and result.get("errcode") and result["errcode"] != 0:
            raise WeChatAPIError(
                result.get("errmsg", "Unknown error"),
                result["errcode"],
                result,
            )

        return result

    def get_access_token(self, force_refresh: bool = False) -> str:
        """Get access token, using cached if valid.

        Args:
            force_refresh: Force refresh token even if current one is valid.

        Returns:
            Access token string.

        Raises:
            WeChatAPIError: If API call fails.
        """
        # Return cached token if still valid (with 60s buffer)
        if not force_refresh and self._access_token and time.time() < self._token_expires_at - 60:
            return self._access_token

        result = self._request(
            "GET",
            "/token",
            params={
                "grant_type": "client_credential",
                "appid": self.app_id,
                "secret": self.app_secret,
            },
        )

        self._access_token = result["access_token"]
        self._token_expires_at = time.time() + result.get("expires_in", 7200)

        return self._access_token

    def upload_image(self, image_path: str) -> Dict[str, Any]:
        """Upload image to WeChat server (for article content).

        Args:
            image_path: Local path to the image file.

        Returns:
            Dict with media_id and url.
        """
        token = self.get_access_token()
        url = f"{self.BASE_URL}/media/uploadimg"
        params = {"access_token": token, "type": "image"}

        with open(image_path, "rb") as f:
            files = {"media": (image_path.split("/")[-1], f, "image/png")}
            resp = requests.post(
                url, params=params, files=files, proxies=self._get_proxies(), timeout=60
            )

        result = resp.json()
        if result.get("errcode") and result["errcode"] != 0:
            raise WeChatAPIError(
                result.get("errmsg", "Upload failed"),
                result["errcode"],
                result,
            )

        return result

    def upload_thumb(self, thumb_path: str) -> str:
        """Upload thumb/cover image, return media_id.

        Args:
            thumb_path: Local path to the thumb image file.

        Returns:
            media_id string.
        """
        token = self.get_access_token()
        url = f"{self.BASE_URL}/material/add_material"
        params = {"access_token": token, "type": "thumb"}

        with open(thumb_path, "rb") as f:
            files = {"media": (thumb_path.split("/")[-1], f, "image/jpeg")}
            resp = requests.post(
                url, params=params, files=files, proxies=self._get_proxies(), timeout=60
            )

        result = resp.json()
        if result.get("errcode") and result["errcode"] != 0:
            raise WeChatAPIError(
                result.get("errmsg", "Upload thumb failed"),
                result["errcode"],
                result,
            )

        return result["media_id"]

    def add_draft(
        self,
        title: str,
        content: str,
        author: str = None,
        thumb_media_id: str = None,
        digest: str = None,
    ) -> Dict[str, Any]:
        """Create a draft article.

        Args:
            title: Article title (max 64 chars).
            content: Article content in HTML format.
            author: Author name (max 8 chars).
            thumb_media_id: Cover image media_id.
            digest: Article summary (max 54 chars).

        Returns:
            Dict with media_id of created draft.
        """
        token = self.get_access_token()
        url = f"{self.BASE_URL}/draft/add"
        params = {"access_token": token}

        articles = [
            {
                "title": title,
                "author": author or "",
                "digest": digest or "",
                "content": content,
                "thumb_media_id": thumb_media_id or "",
                "need_open_comment": 1,
                "only_fans_can_comment": 0,
            }
        ]

        # Use ensure_ascii=False to send UTF-8 directly (avoids Unicode escape issues)
        import json

        json_str = json.dumps({"articles": articles}, ensure_ascii=False)
        result = self._request(
            "POST",
            "/draft/add",
            params=params,
            data=json_str.encode("utf-8"),
        )

        return result

    def get_draft_count(self) -> int:
        """Get total count of drafts."""
        token = self.get_access_token()
        result = self._request(
            "GET",
            "/draft/count",
            params={"access_token": token},
        )
        return result.get("total_count", 0)

    def get_draft_list(self, offset: int = 0, count: int = 20) -> Dict[str, Any]:
        """Get draft list.

        Args:
            offset: Offset for pagination.
            count: Number of items to return (max 20).

        Returns:
            Dict with item list and total count.
        """
        token = self.get_access_token()
        result = self._request(
            "POST",
            "/draft/batchget",
            params={"access_token": token},
            json_data={"offset": offset, "count": count, "no_content": 0},
        )
        return result

    def get_draft(self, media_id: str) -> Dict[str, Any]:
        """Get single draft detail.

        Args:
            media_id: The draft media_id.

        Returns:
            Draft content.
        """
        token = self.get_access_token()
        result = self._request(
            "POST",
            "/draft/get",
            params={"access_token": token},
            json_data={"media_id": media_id},
        )
        return result

    def delete_draft(self, media_id: str) -> Dict[str, Any]:
        """Delete a draft.

        Args:
            media_id: The draft media_id to delete.

        Returns:
            Success status.
        """
        token = self.get_access_token()
        result = self._request(
            "POST",
            "/draft/delete",
            params={"access_token": token},
            json_data={"media_id": media_id},
        )
        return result

    def publish_draft(self, media_id: str) -> Dict[str, Any]:
        """Publish a draft (submit for review).

        Note: Only available for certified accounts.

        Args:
            media_id: The draft media_id to publish.

        Returns:
            Publish result.
        """
        token = self.get_access_token()
        result = self._request(
            "POST",
            "/freepublish/submit",
            params={"access_token": token},
            json_data={"media_id": media_id},
        )
        return result

    def batch_get_material(
        self, type_: str = "image", offset: int = 0, count: int = 20
    ) -> List[Dict]:
        """Get permanent materials list.

        Args:
            type_: Material type: "image", "video", "voice", "news"
            offset: Offset for pagination.
            count: Number of items to return (max 20).

        Returns:
            List of materials.
        """
        token = self.get_access_token()
        result = self._request(
            "POST",
            "/material/batchget_material",
            params={"access_token": token},
            json_data={"type": type_, "offset": offset, "count": count},
        )
        return result.get("item", [])


# Singleton instance
_api: Optional[WeChatAPI] = None


def get_api() -> WeChatAPI:
    """Get global API instance."""
    global _api
    if _api is None:
        _api = WeChatAPI()
    return _api
